gcov: Add a convenient way to generate code coverage report

honggyukim opened the pull request:
This PR provides a way to generate code coverage report using gcov and lcov.
It'd be useful to see the code coverage of uftrace itself for debugging purposes.
It also includes some scripts from lcov project.

The execution sequence is as follows:

1. build uftrace with gcov enabled
  $ make GCOV=1

2. run uftrace
  $ ./uftrace info

3. run a script to generate html report
  $ ./misc/run-lcov.sh

4. open `uftrace.html` in a browser

honggyukim commented:
Sample snapshot of the code coverage report is as follows:

**1. High level overview report**
![image](https://user-images.githubusercontent.com/19642328/41141854-7d3c0770-6b2e-11e8-984d-6c7116cd53c9.png)

**2. line coverage report for a single file**

![image](https://user-images.githubusercontent.com/19642328/41141897-bbc0dc78-6b2e-11e8-838f-5775b5b7c7ef.png)

namhyung commented on Makefile:
```diff
@@ -319,6 +319,10 @@ clean:
+gcov-clean:
+	$(Q)find -name "*\.gcda" | xargs $(RM)
+	$(Q)$(RM) coverage.info
+
```
Then why not just `make clean` having dependency of `make gcov-clean` in the previous commit?

honggyukim replied:
It's because `make gcov-clean` removes only `.gcda` files so that another execution can generate coverage info normally. If `.gcno` files are also removed, we have to recompile uftrace.
And I may have to change this rule to `make coverage-clean` if you think it's better.

namhyung commented on Makefile:
```diff
@@ -86,10 +86,18 @@ endif
   # cannot add -pg to LIB_CFLAGS because mcount() is not reentrant
 endif
 
+ifeq ($(GCOV), 1)
```
Instead of using name of a specific tool, what about a generic name like 'COVERAGE' (provided that it won't called frequently)?

honggyukim replied:
Okay. I will change it to `COVERAGE`.

honggyukim commented:
@namhyung I just updated as you said and I also added branch coverage info. So please test it without yourself again then let me know your feedback. Thanks!

namhyung commented:
I see following warnings.. it seems you also need to turn off the `_FORTIFY_SOURCE`.

```
In file included from /usr/include/bits/libc-header-start.h:33,
                 from /usr/include/stdio.h:27
                 from /home/namhyung/project/uftrace/cmds/record.c:1
/usr/include/features.h:381:4: warning: #warning _FORTIFY_SOURCE requires compiling with optimization (-O) [-Wcpp]
```

honggyukim commented:
Like this?
```diff
diff --git a/Makefile b/Makefile
index 993b01b..f87c308 100644
--- a/Makefile
+++ b/Makefile
@@ -92,10 +92,10 @@ ifeq ($(TRACE), 1)
 endif

 ifeq ($(COVERAGE), 1)
-  UFTRACE_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage
-  DEMANGLER_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage
-  SYMBOLS_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage
-  TRACEEVENT_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage
+  UFTRACE_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage -D_FORTIFY_SOURCE=0
+  DEMANGLER_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage -D_FORTIFY_SOURCE=0
+  SYMBOLS_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage -D_FORTIFY_SOURCE=0
+  TRACEEVENT_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage -D_FORTIFY_SOURCE=0
 endif

 export UFTRACE_CFLAGS LIB_CFLAGS
diff --git a/tests/Makefile b/tests/Makefile
index a5ef862..ec8caef 100644
--- a/tests/Makefile
+++ b/tests/Makefile
@@ -3,7 +3,7 @@ TEST_CFLAGS  += -include $(srcdir)/tests/unittest.h
 TEST_LDFLAGS := -L$(objdir)/libtraceevent -ltraceevent -lelf -pthread -lrt -ldl

 ifeq ($(COVERAGE), 1)
-  TEST_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage
+  TEST_CFLAGS += -O0 -g -fprofile-arcs -ftest-coverage -D_FORTIFY_SOURCE=0
 endif

 UNIT_TEST_SRC := $(srcdir)/uftrace.c
```

namhyung commented:
Hmm.. it generates another warning that it's redefined.  I think you'd be better undefine it like `-U_FORTIFY_SOURCE`.

honggyukim commented:
Thanks. I just updated it using `-U_FORTIFY_SOURCE`.
